Add chp. 3 notebook

sarafrank95 opened the pull request:
Hi @sbillinge,

Here is my solution to Chp. 3. 
I created the notebook in a new folder 'JupyterNotebook' under solutions. I can see that the code is a bit harder to read as a human because the cell information is specified as well which maybe can make the review process a bit more difficult. Before line 516 it is mainly Songsheng's helper functions (and some comments from me introducing them) after that it is my code.

sbillinge reviewed (commented):
I think it is easier to edit the notebook after merging (though, fyi, you can see a preview of the notebook in GitHub by clicking on "view file" under the ... so it is not so hard to priview the notebook).

I don't think it makes sense to have it in a separate folder though, so please can you rename it and delete the subfolder.  We should then think how to name it and do it consistently.  I think something like "<filename>_soln2.ipynb" so it is clear that this is a different solution.  I should probably mention this in the book.  I would then like to add text in a md cell at the top, explaining this.  So text like "This botebook contains a second solution to the problem that makes use of helper functions.  It shows how helper functions can be defined then used to speed up the creation of scripts. It also shows gives a template using the Jupyter Notebook format."

My suggestion would be to put this into all the files in case people don't work through the book from the beginning.  But for now, maybe you could add this and change the file structure, then I wlil merge and we can iterate the notebook text back and forth.  In general, I would like if we could put some more explanation into md cells.  What we can do it do a bit more here, but on later ones, refer people back to this one.  We don't have to describe CMI as this was done in solution 1 and in hte book, so we want the explanation to be about the helper functions and how they work.  Make sure to  say above the cell where the functions are defined that normally they would not be modified, and  the reader doesn't have to go through them.  We could also describe how they could be broken out into py files and imported, also for the educational value?

If you want to have  a crack at any of these more descriptive MD cells, please go aheead, or we can iterate.

sarafrank95 commented:
Okay nice, I can indeed preview when I do as you describe, although I did add some commands to a raw NBConvert cell that does not show up on GH. 
I am unsure if the section about 'Running the notebook' should be included, but I at least had to google around to make the notebook run in the diffpy env. 

I think an important thing to add is where the original helper functions are/cite them.

A lot of what I considered writing in the MD cells right above the functions is already written under the """...""" section in the code. But is it the same information just a bit more readable we want to be stated before the code, so people don't have to look at it?

sarafrank95 commented:
Hi @sbillinge,
I added all the helper functions in a python script: helper_functions.py that is imported at the end of the notebook. I did not repeat all of the solution but just showed how it could be used. 
Right now I put the script in the solutions folder to chp. 3 but I am not sure where it would make the most sense to place it. I also added more comments in the MD cells. Let me know what you think.
